google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,474 @@
+      {
+        std::lock_guard<std::mutex> g{mu_};
+        in_progress_write_->set_value();
+        in_progress_write_ = absl::optional<promise<void>>();
+      }
+
+      if (write_fu.get()) {
+        return make_ready_future(true);
+      }
+
+      bool retrying = false;
+
+      {
+        std::lock_guard<std::mutex> g{mu_};
+        write_reinit_done_.emplace(absl::make_optional(promise<void>()));
+        retrying = stream_state_ == State::kRetrying;
+      }
+
+      if (!retrying) {
+        bool outstanding_read = false;
+        std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>
+            stream;
+        future<void> outstanding_read_future;
+        {
+          std::lock_guard<std::mutex> g{mu_};
+          stream_state_ = State::kRetrying;
+
+          outstanding_read = in_progress_read_.has_value();
+          if (outstanding_read) {
+            outstanding_read_future = in_progress_read_->get_future();
+          }
+
+          stream = std::move(stream_);
+        }
+
+        if (outstanding_read) {
+          outstanding_read_future.then(
+              [this, stream = std::move(stream)](future<void>) {
+                FinishOnStreamFail(std::move(stream));
+              });
+        } else {
+          FinishOnStreamFail(std::move(stream));
+        }
+      }
+
+      return write_reinit_done_->get_future().then(
+          [](future<void>) { return make_ready_future(bool); });
+    });
+  }
+
+  future<Status> Finish() override {
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      if (stream_state_ == State::kShutdown) {
+        return make_ready_future(kPermanentErrorStatus);
+      }
+      stream_state_ = State::kShutdown;
+    }
+
+    // Since `Finish` is being called, there must be no outstanding `Read` and
+    // `Write`. Along with the status not being `State::kShutdown`, this means
+    // that the internal stream is still valid, so we can call `Finish()` on it.
+    return stream_->Finish().then([this](future<Status> finish_future) {
+      auto finish_response = finish_future.get();
+      status_promise_.set_value(finish_response);
+      return make_ready_future(finish_response);
+    });
+  }
+
+ private:
+  enum class State { kRetrying, kInitialized, kShutdown };
+
+  void OnFailure(Status status) {
+    AttemptRetry(status, retry_policy_prototype_->clone(),
+                 backoff_policy_prototype_->clone());
+  }
+
+  void FinishOnStreamFail(
+      std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>
+          stream) {
+    stream->Finish().then([this](future<Status> finish_status) {
+      OnFailure(finish_status.get());
+    });
+  }
+
+  void SetReadWriteFutures() {
+    if (read_reinit_done_.has_value()) {
+      // want to set member variable to nullopt before setting future returned
+      // to caller
+      auto read_promise = read_reinit_done_;
+      read_reinit_done_ = absl::optional();
```

Comment:
you need to hold the lock for modifying the variable, but not hold it for setting the future